src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -83,56 +83,157 @@ ShenandoahMmuTracker::~ShenandoahMmuTracker() {
+      _max_desired_young_size = calculate_max_size(heap_size);
+      break;
+    case SizerNewSizeOnly:
+      _max_desired_young_size = calculate_max_size(heap_size);
+      _max_desired_young_size = MAX2(_min_desired_young_size, _max_desired_young_size);
+      break;
+    case SizerMaxNewSizeOnly:
+      _min_desired_young_size = calculate_min_size(heap_size);
+      _min_desired_young_size = MIN2(_min_desired_young_size, _max_desired_young_size);
+      break;
+    case SizerMaxAndNewSize:
+      // Do nothing. Values set on the command line, don't update them at runtime.
+      break;
+    case SizerNewRatio:
+      _min_desired_young_size = MAX2((heap_size / (NewRatio + 1)), ShenandoahHeapRegion::region_size_bytes());
+      _max_desired_young_size = _min_desired_young_size;
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+
+  assert(_min_desired_young_size <= _max_desired_young_size, "Invalid min/max young gen size values");
+}
+
+void ShenandoahGenerationSizer::heap_size_changed(size_t heap_size) {
+  recalculate_min_max_young_length(heap_size);
+}
+
+bool ShenandoahGenerationSizer::adjust_generation_sizes() {
   shenandoah_assert_generational();
-  if (_mmu_average.davg() >= double(GCTimeRatio)) {
+  if (!use_adaptive_sizing()) {
+    return false;
+  }
+
+  if (_mmu_tracker->average() >= double(GCTimeRatio)) {
     return false;
   }
 
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   ShenandoahOldGeneration *old = heap->old_generation();
-  double old_time_s = old->reset_collection_time();
   ShenandoahYoungGeneration *young = heap->young_generation();
-  double young_time_s = young->reset_collection_time();
   ShenandoahGeneration *global = heap->global_generation();
+  double old_time_s = old->reset_collection_time();
+  double young_time_s = young->reset_collection_time();
   double global_time_s = global->reset_collection_time();
 
+  const double transfer_threshold = 3.0;
+  double delta = young_time_s - old_time_s;
+
   log_info(gc)("Thread Usr+Sys YOUNG = %.3f, OLD = %.3f, GLOBAL = %.3f", young_time_s, old_time_s, global_time_s);
 
-  if (old_time_s > young_time_s) {
-    return transfer_capacity(young, old);
-  } else {
-    return transfer_capacity(old, young);
+  if (abs(delta) <= transfer_threshold) {
+    log_info(gc, ergo)("Difference (%.3f) for thread utilization for each generation is under threshold (%.3f)", abs(delta), transfer_threshold);
+    return false;
   }
-}
 
-size_t percentage_of_heap(size_t bytes) {
-  size_t heap_capacity = ShenandoahHeap::heap()->max_capacity();
-  assert(bytes < heap_capacity, "Must be less than total capacity");
-  return size_t(100.0 * double(bytes) / double(heap_capacity));
+  if (delta > 0) {
+    // young is busier than old, increase size of young to raise MMU
+    return transfer_capacity(old, young);
+  } else {
+    // old is busier than young, increase size of old to raise MMU
+    return transfer_capacity(young, old);
+  }
 }
 
```

Comment:
I still question whether this is the right trigger for when to enlarge old-gen.  A properly running generational GC will spend nearly all its time doing young-gen and very little time doing old-gen.

The trigger for enlarging old-gen should be that we experience promotion failures (and/or that we identify at the end of init mark that we have more live data in aged regions than will fit in the current old-gen).

Old-gen collection triggers need to be refined when we are auto-sizing.  We can't use "percent free in old" or even "time to collect old > time to exhaust old", because we are trying to auto-tune to maintain that the percent free in old is very small.  We need a new way to trigger old-gen GCs.

Maybe, we trigger old-gen GC (rather than enlarging old-gen) if an old-gen enlargement request would cause us to exceed a target max-size for old-gen?

This doesn't have to be addressed in current PR.  But I'm just thinking that refinements of these heuristics may be necessary eventually.

Reply:
Maybe the target max-size for old-gen is auto-tuned to, e.g., 10% larger than the maximum old-gen live memory following old-gen concurrent mark with some bias given to more recently observed measurements of old-gen live-memory.

Reply:
I agree we can wire up more signals to the resizing mechanism. In the scenario you describe, where old generation has become _too small_ and old collections are running _too frequently_, the MMU based resizing would enlarge the old generation.

Reply:
What do you do before any data is available for one of the MMU trackers? (for example before the first old collection cycle has happened.) I'd assume that the control algorithm wouldn't kick in until the data driving the control signal was valid and available. Where is that done?

Reply:
We generally have a flurry of "learning cycles" at startup, but it makes sense to add an explicit check to avoid premature resizing. To Kelvin's earlier point, it may also make sense to re-initiate the learning phase after resizing the generations.